Index: Source/NSFileManager.m
===================================================================
--- Source/NSFileManager.m	(revision 28238)
+++ Source/NSFileManager.m	(working copy)
@@ -388,6 +388,7 @@
 - (BOOL) changeFileAttributes: (NSDictionary*)attributes atPath: (NSString*)path
 {
   const _CHAR	*lpath = 0;
+  NSNumber	*tmpNum;
   unsigned long	num;
   NSString	*str;
   NSDate	*date;
@@ -400,7 +401,8 @@
   lpath = [defaultManager fileSystemRepresentationWithPath: path];
 
 #ifndef __MINGW32__
-  num = [attributes fileOwnerAccountID];
+  tmpNum = [attributes fileOwnerAccountID];
+  num = tmpNum ? [tmpNum unsignedLongValue] : NSNotFound;
   if (num != NSNotFound)
     {
       if (chown(lpath, num, -1) != 0)
@@ -455,7 +457,8 @@
 	}
     }
 
-  num = [attributes fileGroupOwnerAccountID];
+  tmpNum = [attributes fileGroupOwnerAccountID];
+  num = tmpNum ? [tmpNum unsignedLongValue] : NSNotFound;
   if (num != NSNotFound)
     {
       if (chown(lpath, -1, num) != 0)
@@ -769,7 +772,7 @@
    * If there is no file owner specified, and we are running setuid to
    * root, then we assume we need to change ownership to correct user.
    */
-  if (attributes == nil || ([attributes fileOwnerAccountID] == NSNotFound
+  if (attributes == nil || ([attributes fileOwnerAccountID] == nil
     && [attributes fileOwnerAccountName] == nil))
     {
       if (geteuid() == 0 && [@"root" isEqualToString: NSUserName()] == NO)
@@ -932,7 +935,7 @@
    * If there is no file owner specified, and we are running setuid to
    * root, then we assume we need to change ownership to correct user.
    */
-  if (attributes == nil || ([attributes fileOwnerAccountID] == NSNotFound
+  if (attributes == nil || ([attributes fileOwnerAccountID] == nil
     && [attributes fileOwnerAccountName] == nil))
     {
       if (geteuid() == 0 && [@"root" isEqualToString: NSUserName()] == NO)
@@ -2401,18 +2404,12 @@
 }
 
 /**
- * Return the numeric value of the NSFileOwnerAccountID attribute
- * in the dictionary, or NSNotFound if the attribute is not present.
+ * Return an NSNumber with the numeric value of the NSFileOwnerAccountID attribute
+ * in the dictionary, or nil if the attribute is not present.
  */
-- (unsigned long) fileOwnerAccountID
+- (NSNumber*) fileOwnerAccountID
 {
-  NSNumber	*n = [self objectForKey: NSFileOwnerAccountID];
-
-  if (n == nil)
-    {
-      return NSNotFound;
-    }
-  return [n unsignedIntValue];
+  return [self objectForKey: NSFileOwnerAccountID];
 }
 
 /**
@@ -2424,18 +2421,12 @@
 }
 
 /**
- * Return the numeric value of the NSFileGroupOwnerAccountID attribute
- * in the dictionary, or NSNotFound if the attribute is not present.
+ * Return an NSNumber with the numeric value of the NSFileGroupOwnerAccountID attribute
+ * in the dictionary, or nil if the attribute is not present.
  */
-- (unsigned long) fileGroupOwnerAccountID
+- (NSNumber*) fileGroupOwnerAccountID
 {
-  NSNumber	*n = [self objectForKey: NSFileGroupOwnerAccountID];
-
-  if (n == nil)
-    {
-      return NSNotFound;
-    }
-  return [n unsignedIntValue];
+  return [self objectForKey: NSFileGroupOwnerAccountID];
 }
 
 /**
Index: Headers/Foundation/NSFileManager.h
===================================================================
--- Headers/Foundation/NSFileManager.h	(revision 28238)
+++ Headers/Foundation/NSFileManager.h	(working copy)
@@ -460,9 +460,9 @@
 - (BOOL) fileIsImmutable;
 - (unsigned long long) fileSize;
 - (NSString*) fileType;
-- (unsigned long) fileOwnerAccountID;
+- (NSNumber*) fileOwnerAccountID;
 - (NSString*) fileOwnerAccountName;
-- (unsigned long) fileGroupOwnerAccountID;
+- (NSNumber*) fileGroupOwnerAccountID;
 - (NSString*) fileGroupOwnerAccountName;
 - (NSDate*) fileModificationDate;
 - (NSUInteger) filePosixPermissions;
